# Task 3

First, update the following cell to have a link to your Homework 1 GitHub repository:



**UPDATE THIS LINK:**

[https://github.com/carmengg/eds-220-hwk-1]()



In this task, make at least 5 commits at moments you think are adequate. We'll look at your repo to check commit history.

Add comments for all your code. Err on the side of commenting too much for now. 

Comments in each exercise and commits will count towards your grade.

This task covers topics we will see on Monday Oct. 9. The best strategy to solve it is to get together with your classmates and work on it together after Monday's lecture.


## 1
For this task we are going to use data about  [Western Indian Ocean Coral Diversity (McClanahan, 2023)](https://knb.ecoinformatics.org/view/doi%3A10.5063%2FF1K35S3H) from the Knowledge Network for Biocomplexity (KNB) data repository. 

a. Navigate to the dataset's link and explore the data package. 
 
b. Use the next markdown cell to add a brief description of the dataset, including a citation, date of access, and a link to the archive. 


**This dataset is based on coral taxa in Africa off the coast. The habitats of each coral vary and according to the paper, they used two prediction models (one for 2020 and the other for 2050) to measure multiple temperature and non-temperature variables.**

**Tim McClanahan. (2023). Western Indian Ocean Coral Diversity. Knowledge Network for Biocomplexity. doi:10.5063/F1K35S3H.**

**DOA: 10/10/23**

**https://knb.ecoinformatics.org/knb/d1/mn/v2/object/doi%3A10.5063%2FF1K35S3H**

## 2
In the next code cell, import the pandas package using standard abbreviation in a code cell.

Read in the data from the URL using the `pd.read_csv` function and store it as `coral_div` like this:

```
# read in data
coral_div = pd.read_csv('the URL goes here')
```

In [6]:
#import library
import pandas as pd

#read in the data 

coral_div = pd.read_csv('https://knb.ecoinformatics.org/knb/d1/mn/v2/object/urn%3Auuid%3Ae7d12a5a-f1a9-4a3e-aaa5-5222f67e799d')


coral_div

PU_ID     PU_Lat     PU_Lon  Number_of_genera       Country  \
0        7  39.076711  -5.490011                35      Tanzania   
1        7  39.076711  -5.490011                22      Tanzania   
2        7  39.076711  -5.490011                25      Tanzania   
3        7  39.076711  -5.490011                26      Tanzania   
4       10  39.054253  -5.512514                27      Tanzania   
..     ...        ...        ...               ...           ...   
570  11681  32.878336 -27.011417                19  South Africa   
571  11682  32.878336 -27.011417                18  South Africa   
572  11683  32.878336 -27.011417                19  South Africa   
573  11683  32.878336 -27.011417                19  South Africa   
574  11683  32.878336 -27.011417                21  South Africa   

                    Ecoregion   Management           Habitat Observer  Depth  \
0    East African Coral Coast  UnfishedLow             Crest      Tim  11.00   
1    East African Coral Coast  UnfishedLow  Lagoon/Back reef  Nyawira   2.00   
2    East African Coral Coast  UnfishedLow             Slope  Nyawira   4.00   
3    East African Coral Coast  UnfishedLow             Slope  Nyawira   3.00   
4    East African Coral Coast  UnfishedLow  Lagoon/Back reef  Nyawira   1.50   
..                        ...          ...               ...      ...    ...   
570                   Delagoa   Restricted             Slope     Sink  15.50   
571                   Delagoa   Restricted             Slope   Carlos  12.00   
572                   Delagoa   Restricted             Slope   Carlos  12.60   
573                   Delagoa   Restricted             Slope   Porter  13.00   
574                   Delagoa   Restricted             Slope     Sink  13.47   

     ...  SST_bimodality  SST_Skewness  ClimateStressModel   Netflow  \
0    ...        0.462556     -0.128136            0.596375  0.144498   
1    ...        0.462556     -0.128136            0.596375  0.144498   
2    ...        0.462556     -0.128136            0.596375  0.144498   
3    ...        0.462556     -0.128136            0.596375  0.144498   
4    ...        0.461297     -0.134548            0.629000  0.011721   
..   ...             ...           ...                 ...       ...   
570  ...        0.568628      0.084750            0.371750  0.030092   
571  ...        0.568628      0.084750            0.371750  0.030092   
572  ...        0.568628      0.084750            0.371750  0.030092   
573  ...        0.568628      0.084750            0.371750  0.030092   
574  ...        0.568628      0.084750            0.371750  0.030092   

     Indegree  Outdegree  Retention  min_yr  max_yr  no_replicates  
0       98.25      74.75      14.00    2019    2019              1  
1       98.25      74.75      14.00    2010    2015              2  
2       98.25      74.75      14.00    2015    2015              1  
3       98.25      74.75      14.00    2016    2019              2  
4      112.75      51.50     183.75    2010    2019              3  
..        ...        ...        ...     ...     ...            ...  
570     15.75      17.00      86.50    2015    2016              2  
571     15.75      17.00      86.50    2007    2007              2  
572     15.75      17.00      86.50    2005    2007             10  
573     15.75      17.00      86.50    2022    2022              4  
574     15.75      17.00      86.50    2015    2017             15  

[575 rows x 42 columns]

## 3

Use attributes and methods of the `coral_div` data frame to obtain the following information about it:

a. how many rows and columns does it have?

b. what are the column names?

c. what are the data types of the columns?

d. display the first 8 rows with all the column names. For this one, you will need to udpate a `pandas` option (see discussion section)

e. some inital info you are interested in. 

Use the code cells marked a-e. For each, add a comment explaining what is the ouput.

In [11]:
# a
coral_div.shape #use the shape function to show how many rows and cols there are 

(575, 42)

**There are 575 rows in coral_div data frame and 42 columns.**

In [13]:
# b

coral_div.columns #print out the column names 

In [17]:
# c

coral_div.dtypes #use the function to find data type of each col

PU_ID                    int64
PU_Lat                 float64
PU_Lon                 float64
Number_of_genera         int64
Country                 object
Ecoregion               object
Management              object
Habitat                 object
Observer                object
Depth                  float64
Coralcover               int64
SSTRateofRise          float64
PARmax                 float64
Calcite                float64
Dis_oxygen             float64
mean.waveNRJ           float64
Salinity_mean          float64
andrello_reef_value      int64
andrello_nutrients     float64
mean.npp               float64
TT_market_hrs          float64
TT_pop_hrs             float64
cumDHW                 float64
SST_Kurtosis           float64
SST_Median             float64
ChlorA_median          float64
andrello_sediments     float64
Current_vel_mean       float64
Diff_attenuation       float64
Grav_NP                float64
Grav_NC                float64
PH                     float64
SST_bimo

In [19]:
# d

coral_div.head(8) #in the paranthesis put number of rows you want printed 

PU_ID     PU_Lat    PU_Lon  Number_of_genera   Country  \
0      7  39.076711 -5.490011                35  Tanzania   
1      7  39.076711 -5.490011                22  Tanzania   
2      7  39.076711 -5.490011                25  Tanzania   
3      7  39.076711 -5.490011                26  Tanzania   
4     10  39.054253 -5.512514                27  Tanzania   
5     10  39.054253 -5.512514                29  Tanzania   
6     15  39.054253 -5.535017                19  Tanzania   
7     39  39.009337 -5.647518                28  Tanzania   

                  Ecoregion   Management           Habitat Observer  Depth  \
0  East African Coral Coast  UnfishedLow             Crest      Tim   11.0   
1  East African Coral Coast  UnfishedLow  Lagoon/Back reef  Nyawira    2.0   
2  East African Coral Coast  UnfishedLow             Slope  Nyawira    4.0   
3  East African Coral Coast  UnfishedLow             Slope  Nyawira    3.0   
4  East African Coral Coast  UnfishedLow  Lagoon/Back reef  Nyawira    1.5   
5  East African Coral Coast  UnfishedLow  Lagoon/Back reef      Tim    1.5   
6  East African Coral Coast   Restricted             Crest      Tim    8.0   
7  East African Coral Coast  UnfishedLow  Lagoon/Back reef  Nyawira    3.0   

   ...  SST_bimodality  SST_Skewness  ClimateStressModel   Netflow  Indegree  \
0  ...        0.462556     -0.128136            0.596375  0.144498     98.25   
1  ...        0.462556     -0.128136            0.596375  0.144498     98.25   
2  ...        0.462556     -0.128136            0.596375  0.144498     98.25   
3  ...        0.462556     -0.128136            0.596375  0.144498     98.25   
4  ...        0.461297     -0.134548            0.629000  0.011721    112.75   
5  ...        0.461297     -0.134548            0.629000  0.011721    112.75   
6  ...        0.461297     -0.134548            0.595600  0.011721    112.75   
7  ...        0.464874     -0.158601            0.626000 -0.063087     96.50   

   Outdegree  Retention  min_yr  max_yr  no_replicates  
0      74.75      14.00    2019    2019              1  
1      74.75      14.00    2010    2015              2  
2      74.75      14.00    2015    2015              1  
3      74.75      14.00    2016    2019              2  
4      51.50     183.75    2010    2019              3  
5      51.50     183.75    2015    2015              1  
6      51.50     183.75    2019    2019              1  
7      40.75      35.75    2015    2019              3  

[8 rows x 42 columns]

In [32]:
# e

# select a single column by using square brackets []
country_name = coral_div['Country']

# check the type of the ouput
print(type(country_name))

#Look at the first 5 rows of col country 
country_name.head()


<class 'pandas.core.series.Series'>


0    Tanzania
1    Tanzania
2    Tanzania
3    Tanzania
4    Tanzania
Name: Country, dtype: object

## 4

Some of the columns have data type `int64` and some have data type `float64`. What is the difference between these `int64` and `float64`?


**`float64` is a type of numeric value with decimals. `int64` is a numeric whole value.**

## 5

The `Country` column indicates the country of reef location. Explain in your words what is the ouput of `coral_div.Country` and `coral_div.Country.unique()`. 

In [33]:
#use this code to outline which cols 
coral_div.Country

0          Tanzania
1          Tanzania
2          Tanzania
3          Tanzania
4          Tanzania
           ...     
570    South Africa
571    South Africa
572    South Africa
573    South Africa
574    South Africa
Name: Country, Length: 575, dtype: object

The output of `coral_div.Country` is the first five rows and last five rows of the country column in the coral_div data set. It also shows the length of the column which is 575 rows, and the type of the column which is `object`.

In [34]:
coral_div.Country.unique()

array(['Tanzania', 'Mozambique', 'Kenya', 'Reunion', 'Comoros', 'Mayotte',
       'French Southern Territories', 'Madagascar', 'Seychelles',
       'South Africa', 'Mauritius'], dtype=object)

The code `coral_div.Country.unique()` shows the unique objects in the `Country` columns. Each one listed is shown at least once in column.

## 6

Check the dataset's [metadata file README_md.doc](https://knb.ecoinformatics.org/view/doi:10.5063/F1K35S3H) to find which column has information about who made an observation. How many observers collected this data? HINT: use the `len` Python function. Not sure what `len` does? Use `?`.

In [38]:
# solve exercise here

#use unique function to pick out each name listed at each once
len(coral_div.Observer.unique()) #use len function to call the number of observers

16

The total number of observers who collected the data are 16.

## 7

The `Coralcover` column indicates the percent of hard coral cover in each 6km reef cell surveyed. What data is `coral_div[coral_div['Coralcover'] >= 50]` selecting?

In [46]:
#did this to check the first couple columns covered
coral_div.Coralcover.head()

coral_div[coral_div['Coralcover'] >= 50]

PU_ID     PU_Lat     PU_Lon  Number_of_genera     Country  \
39     220  39.233916  -6.658981                17    Tanzania   
40     220  39.233916  -6.658981                19    Tanzania   
41     221  39.256374  -6.658981                25    Tanzania   
42     221  39.256374  -6.658981                19    Tanzania   
43     221  39.256374  -6.658981                12    Tanzania   
..     ...        ...        ...               ...         ...   
303   4039  48.082322 -13.553193                27  Madagascar   
304   4039  48.082322 -13.553193                20  Madagascar   
305   4039  48.082322 -13.553193                21  Madagascar   
306   4039  48.082322 -13.553193                33  Madagascar   
514   6064  40.603847 -12.496246                30  Mozambique   

                           Ecoregion   Management           Habitat Observer  \
39          East African Coral Coast  UnfishedLow             Slope  Nyawira   
40          East African Coral Coast  UnfishedLow             Slope      Tim   
41          East African Coral Coast  UnfishedLow  Lagoon/Back reef  Nyawira   
42          East African Coral Coast  UnfishedLow  Lagoon/Back reef      Tim   
43          East African Coral Coast  UnfishedLow  Lagoon/Back reef  Nyawira   
..                               ...          ...               ...      ...   
303  Western and Northern Madagascar  UnfishedLow             Slope  Arielle   
304  Western and Northern Madagascar  UnfishedLow             Slope  D'Agata   
305  Western and Northern Madagascar  UnfishedLow             Slope   Julien   
306  Western and Northern Madagascar  UnfishedLow             Slope  Nyawira   
514         East African Coral Coast  UnfishedLow  Lagoon/Back reef  Nyawira   

         Depth  ...  SST_bimodality  SST_Skewness  ClimateStressModel  \
39    3.000000  ...        0.527900     -0.171423             0.62130   
40    2.833333  ...        0.527900     -0.171423             0.62130   
41   10.000000  ...        0.526911     -0.169093             0.52760   
42   10.000000  ...        0.526911     -0.169093             0.52760   
43    3.000000  ...        0.526911     -0.169093             0.52760   
..         ...  ...             ...           ...                 ...   
303   6.733333  ...        0.566251     -0.271831             0.72960   
304   6.000000  ...        0.566251     -0.271831             0.72960   
305   5.000000  ...        0.566251     -0.271831             0.72960   
306   5.000000  ...        0.566251     -0.271831             0.72960   
514   8.000000  ...        0.529942     -0.145626             0.68105   

      Netflow  Indegree  Outdegree  Retention  min_yr  max_yr  no_replicates  
39  -0.443779    127.50      62.25     204.75    2016    2016              1  
40  -0.443779    127.50      62.25     204.75    2012    2019              3  
41  -0.443779    127.50      62.25     204.75    2016    2016              1  
42  -0.443779    127.50      62.25     204.75    2016    2016              1  
43  -0.443779    127.50      62.25     204.75    2016    2016              1  
..        ...       ...        ...        ...     ...     ...            ...  
303  0.670117     12.50      29.00      62.50    2017    2019              3  
304  0.670117     12.50      29.00      62.50    2016    2016              2  
305  0.670117     12.50      29.00      62.50    2016    2016              1  
306  0.670117     12.50      29.00      62.50    2015    2015              1  
514  0.116857     54.75      54.25     126.50    2015    2015              1  

[67 rows x 42 columns]

The code `coral_div[coral_div['Coralcover'] >= 50]` shows all rows that have greater or equal to 50%  hard coral cover in each 6km reef cell surveyed. 

## 8

Use code to find if there are there any sites with a 30% coral cover in Tanzania.

In [143]:
# solve exercise here

coral_div.loc[coral_div['Coralcover'] != 30]

PU_ID     PU_Lat     PU_Lon  Number_of_genera       Country  \
0        7  39.076711  -5.490011                35      Tanzania   
1        7  39.076711  -5.490011                22      Tanzania   
2        7  39.076711  -5.490011                25      Tanzania   
3        7  39.076711  -5.490011                26      Tanzania   
4       10  39.054253  -5.512514                27      Tanzania   
..     ...        ...        ...               ...           ...   
570  11681  32.878336 -27.011417                19  South Africa   
571  11682  32.878336 -27.011417                18  South Africa   
572  11683  32.878336 -27.011417                19  South Africa   
573  11683  32.878336 -27.011417                19  South Africa   
574  11683  32.878336 -27.011417                21  South Africa   

                    Ecoregion   Management           Habitat Observer  Depth  \
0    East African Coral Coast  UnfishedLow             Crest      Tim  11.00   
1    East African Coral Coast  UnfishedLow  Lagoon/Back reef  Nyawira   2.00   
2    East African Coral Coast  UnfishedLow             Slope  Nyawira   4.00   
3    East African Coral Coast  UnfishedLow             Slope  Nyawira   3.00   
4    East African Coral Coast  UnfishedLow  Lagoon/Back reef  Nyawira   1.50   
..                        ...          ...               ...      ...    ...   
570                   Delagoa   Restricted             Slope     Sink  15.50   
571                   Delagoa   Restricted             Slope   Carlos  12.00   
572                   Delagoa   Restricted             Slope   Carlos  12.60   
573                   Delagoa   Restricted             Slope   Porter  13.00   
574                   Delagoa   Restricted             Slope     Sink  13.47   

     ...  SST_bimodality  SST_Skewness  ClimateStressModel   Netflow  \
0    ...        0.462556     -0.128136            0.596375  0.144498   
1    ...        0.462556     -0.128136            0.596375  0.144498   
2    ...        0.462556     -0.128136            0.596375  0.144498   
3    ...        0.462556     -0.128136            0.596375  0.144498   
4    ...        0.461297     -0.134548            0.629000  0.011721   
..   ...             ...           ...                 ...       ...   
570  ...        0.568628      0.084750            0.371750  0.030092   
571  ...        0.568628      0.084750            0.371750  0.030092   
572  ...        0.568628      0.084750            0.371750  0.030092   
573  ...        0.568628      0.084750            0.371750  0.030092   
574  ...        0.568628      0.084750            0.371750  0.030092   

     Indegree  Outdegree  Retention  min_yr  max_yr  no_replicates  
0       98.25      74.75      14.00    2019    2019              1  
1       98.25      74.75      14.00    2010    2015              2  
2       98.25      74.75      14.00    2015    2015              1  
3       98.25      74.75      14.00    2016    2019              2  
4      112.75      51.50     183.75    2010    2019              3  
..        ...        ...        ...     ...     ...            ...  
570     15.75      17.00      86.50    2015    2016              2  
571     15.75      17.00      86.50    2007    2007              2  
572     15.75      17.00      86.50    2005    2007             10  
573     15.75      17.00      86.50    2022    2022              4  
574     15.75      17.00      86.50    2015    2017             15  

[558 rows x 42 columns]

## 9

We are interested in looking at the PH of sea water (`PH` column),  mean salinity (PSS) (`Salinity_mean` column), and depth of survey location in meters (`Depth` column). Select this data. 

In [144]:
# solve exercise here

coral_div[['PH', 'Salinity_mean', 'Depth']] #double brackets to call list

PH  Salinity_mean  Depth
0    8.2175      35.167646  11.00
1    8.2175      35.167646   2.00
2    8.2175      35.167646   4.00
3    8.2175      35.167646   3.00
4    8.2180      35.102688   1.50
..      ...            ...    ...
570  8.2400      35.301208  15.50
571  8.2400      35.301208  12.00
572  8.2400      35.301208  12.60
573  8.2400      35.301208  13.00
574  8.2400      35.301208  13.47

[575 rows x 3 columns]

## 10
The `max_yr` column indicates the year of last sampling at a given site. Select the PH, salinity, and depth for sites where sampling ended in 2017. 


In [153]:
# solve exercise here

max2017 = coral_div[['PH', 'Salinity_mean', 'Depth', 'max_yr']]

max2017[max2017['max_yr'] != 2017]

PH  Salinity_mean  Depth  max_yr
0    8.2175      35.167646   11.0    2019
1    8.2175      35.167646    2.0    2015
2    8.2175      35.167646    4.0    2015
3    8.2175      35.167646    3.0    2019
4    8.2180      35.102688    1.5    2019
..      ...            ...    ...     ...
568  8.2400      35.301208   10.8    2022
570  8.2400      35.301208   15.5    2016
571  8.2400      35.301208   12.0    2007
572  8.2400      35.301208   12.6    2007
573  8.2400      35.301208   13.0    2022

[540 rows x 4 columns]

## 11
Use `iloc` to select rows 95 throuhg 100 (including 100) in the 17th, 20th, and 28th columns and assign these values to a new variable called `subset`. HINT: the ouput columns should be `Salinity_mean`, `mean.npp` and `Current_vel_mean`. Are you indexing from 0?

In [11]:
# solve exercise here

## 12 

Run `subset.isna()`. Explain what this code does and what the output represents.

*Update this markdown cell with your answer.*

## 13
Run `subset.isna().any()`. Explain what this code does and what the output represents. HINT: It could be helpful to check the [documentation of `pandas.Sries.any`](https://pandas.pydata.org/docs/reference/api/pandas.Series.any.html).


*Update this markdown cell with your answer.*

## 14 

Select the rows in `subset` that have NA in the `Current_vel_mean` column. HINT: you can use your answer for 9.

In [12]:
# solve exercise here